File: NotesUIKit/NotesUIKit/Data store/DataStore.swift
//
//  DataStore.swift
//  NotesUIKit
//
//  Created by Andrei Sadovnicov on 20.9.2022.
//

import Foundation
import RealmSwift

class DataStore {
    
    // MARK: - PROPERTIES
    
    // MARK: - Singleton
    static let shared = DataStore()
    
    // MARK: - Realm
    private let realm = try! Realm()
    
    // MARK: - METHODS
    
    // MARK: - CREATE
    func addNote() {
        let note = Note(content: Constants.defaultNoteContent)
        try! realm.write {
            realm.add(note)
        }
    }
    
    // MARK: - READ
    func getAllNotes() -> Results<Note> {
        return realm.objects(Note.self).sorted(byKeyPath: Constants.lastUpdateKeyPath, ascending: false)
    }
    
    func getFavoriteNotes() -> Results<Note> {
        return realm.objects(Note.self).where{$0.isFavorite == true}.sorted(byKeyPath: Constants.lastUpdateKeyPath, ascending: false)
    }
    
    // MARK: - UPDATE
    func toggleFavorite(for note: Note) {
        try! realm.write() {
            note.isFavorite.toggle()
        }
    }
    
    func update(_ note: Note, content: String? = nil) {
        try! realm.write {
            note.lastUpdate = Date()
            if let content = content {
                note.content = content
                note.setTitleAndSubtitle()
            }
        }
    }
    
    // MARK: - DELETE
    func delete(_ note: Note) {
        try! realm.write {
            realm.delete(note)
        }
    }
    
}

File: notesflutter/lib/datastore/datastore.dart
import 'package:isar/isar.dart';
import 'package:notesflutter/constants/constants.dart';
import 'package:notesflutter/models/note.dart';

class DataStore {
  // PROPERTIES

  // Singleton
  DataStore._();
  static final shared = DataStore._();

  // Isar
  late Isar isar;

  // METHODS

  // INITIALIZATION
  Future<void> initialize() async {
    isar = await Isar.open([NoteSchema]);
  }

  // CREATE
  Future<void> addNote() async {
    final note = Note()
      ..content = Constants.defaultNoteContent
      ..setTitleAndSubtitle();

    await isar.writeTxn(() async {
      await isar.notes.put(note);
    });
  }

  // READ
  Stream<List<Note>> getAllNotesStream() {
    final notesQuery = isar.notes.where().sortByLastUpdateDesc().build();
    final stream = notesQuery.watch(fireImmediately: true);
    return stream;
  }

  Stream<List<Note>> getFavoriteNotesStream() {
    final notesQuery = isar.notes.where().sortByLastUpdateDesc().build();
    final stream = notesQuery.watch(fireImmediately: true);
    return stream;
  }

  // UPDATE
  Future<void> updateNote({required Note note, String? content}) async {
    note.lastUpdate = DateTime.now();
    if (content != null) {
      note.content = content;
      note.setTitleAndSubtitle();
    }
    await isar.writeTxn(() async {
      isar.notes.put(note);
    });
  }

  Future<void> toggleFavorite({required Note note}) async {
    note.isFavorite = !note.isFavorite;
    await isar.writeTxn(() async {
      isar.notes.put(note);
    });
  }

  // DELETE
  Future<void> deleteNote({required Note note}) async {
    await isar.writeTxn(() async {
      await isar.notes.delete(note.id);
    });
  }
}
